File: pythonLang/binary_search.py
import sys
import time
import os


def binary_search(arr, target):
    """
    Perform binary search on a sorted array of strings.
    Each string is expected to be comma-separated with a number in the first column.
    """
    left, right = 0, len(arr) - 1
    target_val = int(target.split(",")[0])

    while left <= right:
        mid = (left + right) // 2
        mid_val = int(arr[mid].split(",")[0])

        if mid_val == target_val:
            if arr[mid] == target:
                return mid
            else:
                return -1
        elif mid_val < target_val:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def main():
    if len(sys.argv) != 2:
        print("Usage: python binary_search.py <filename>")
        return

    filename = sys.argv[1]

    try:
        with open(filename, "r") as file:
            data = [line.strip() for line in file if line.strip()]

        best_time, worst_time, total_time = float("inf"), 0, 0
        num_tests = len(data) + 1  # includes not-found case

        for target in data:
            start = time.perf_counter()
            binary_search(data, target)
            end = time.perf_counter()
            elapsed = end - start

            best_time = min(best_time, elapsed)
            worst_time = max(worst_time, elapsed)
            total_time += elapsed

        # Add a not-found target
        not_found_target = "99999999,notfound"
        start = time.perf_counter()
        binary_search(data, not_found_target)
        end = time.perf_counter()
        elapsed = end - start

        worst_time = max(worst_time, elapsed)
        total_time += elapsed

        avg_time = total_time / num_tests

        output_path = f"./output/binary_search_{len(data)}.txt"
        os.makedirs(os.path.dirname(output_path), exist_ok=True)

        with open(output_path, "w") as out:
            out.write(f"Best Case (ms): {best_time * 1000:.6f}\n")
            out.write(f"Average Case (ms): {avg_time * 1000:.6f}\n")
            out.write(f"Worst Case (ms): {worst_time * 1000:.6f}\n")

        print(f"Binary search timing written to {output_path}")

    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()

File: ada_assign/code/javaLang/BinarySearch.java
package javaLang;

import java.io.*;
import java.util.*;

public class BinarySearch {
    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Usage: java BinarySearch <filename>");
            return;
        }

        String filename = args[0];
        List<String> data = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    data.add(line.trim());
                }
            }
        }

        int n = data.size();
        long bestTime = Long.MAX_VALUE, worstTime = Long.MIN_VALUE, totalTime = 0;

        for (String target : data) {
            long startTime = System.nanoTime();
            binarySearch(data, target);
            long duration = System.nanoTime() - startTime;

            bestTime = Math.min(bestTime, duration);
            worstTime = Math.max(worstTime, duration);
            totalTime += duration;
        }

        // Add a not-found target
        String notFoundTarget = "99999999,notfound";
        long startTime = System.nanoTime();
        binarySearch(data, notFoundTarget);
        long duration = System.nanoTime() - startTime;

        worstTime = Math.max(worstTime, duration);
        totalTime += duration;
        n++; // include not-found test in average

        // Convert nanoseconds to milliseconds
        double bestMs = bestTime / 1_000_000.0;
        double avgMs = (totalTime / (double) n) / 1_000_000.0;
        double worstMs = worstTime / 1_000_000.0;

        File outputDir = new File("./output");
        outputDir.mkdirs();
        String outputFile = "./output/binary_search_" + (n - 1) + ".txt";

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            writer.write(String.format("Best Case (ms): %.6f\n", bestMs));
            writer.write(String.format("Average Case (ms): %.6f\n", avgMs));
            writer.write(String.format("Worst Case (ms): %.6f\n", worstMs));
        }

        System.out.println("Binary search timing written to " + outputFile);
    }

    static int binarySearch(List<String> arr, String targetEntry) {
        int targetVal = Integer.parseInt(targetEntry.split(",")[0]);
        int left = 0, right = arr.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            String midEntry = arr.get(mid);
            int midVal = Integer.parseInt(midEntry.split(",")[0]);

            if (midVal == targetVal) {
                if (midEntry.equals(targetEntry)) {
                    return mid;
                } else {
                    return -1;
                }
            } else if (midVal < targetVal) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}
